src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/VTableSliceNode.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.TypeSystem;
+using System;
+using System.Collections.Generic;
+using Debug = System.Diagnostics.Debug;
+using ILCompiler.DependencyAnalysisFramework;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents the VTable for a type's slice. For example, System.String's VTableSliceNode includes virtual 
+    /// slots added by System.String itself, System.Object's EETypeVTableSliceNode contains the virtuals it defines.
+    /// </summary>
+    internal class VTableSliceNode : DependencyNodeCore<NodeFactory>
+    {
+        TypeDesc _type;
+        bool _shouldBuildFullVTable;
+        List<MethodDesc> _slots = new List<MethodDesc>();
+
+        public VTableSliceNode(TypeDesc type, NodeFactory context)
+        {
+            _type = type;
+            if (context.CompilationModuleGroup.ShouldProduceFullType(_type))
+            {
+                _shouldBuildFullVTable = true;
+            }
+        }
+        
+        public IReadOnlyList<MethodDesc> Slots
+        {
+            get
+            {
+                return _slots;
```

Comment:
(It should also freeze the VTableSlices of base types to be really effective at killing bugs)